subsys/debug/gdbstub.c

```diff
@@ -549,6 +549,16 @@ static int gdb_mem_write(const uint8_t *buf, uintptr_t addr,
 	}
 
 out:
+#ifdef CONFIG_XTENSA
+	/* This flush sequence was taken from the Xtensa manual under the ihi
+	 * instruction. It ensures breakpoints set by overwriting the instruction
+	 * memory will be hit.
+	 */
+	xthal_dcache_line_writeback((void *)addr);
+	xthal_icache_sync();
+	xthal_icache_line_invalidate((void *)addr);
+	xthal_icache_sync();
```

Comment:
Better to emit the assembly directly.  We're trying to get away from the HAL library.  For dcache, there's already an API for this in https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/arch/xtensa/cache.h you could augment with instruction cache variants.